### PR TITLE
Added Gradle Support using JK1 plugin

### PR DESCRIPTION
**Pull request by chakrapanich**
Hi Team, We have implemented Gradle support using JK1 plugin.

https://github.com/jk1/Gradle-License-Report

### CODE REVIEW

File: src/main/java/at/porscheinformatik/sonarqube/licensecheck/ScannerResolver.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import at.porscheinformatik.sonarqube.licensecheck.mavendependency.MavenDependencyService;
+import at.porscheinformatik.sonarqube.licensecheck.mavenlicense.MavenLicenseService;
+import at.porscheinformatik.sonarqube.licensecheck.npm.PackageJsonDependencyScanner;
+import org.sonar.api.config.Configuration;
+
+import java.io.*;
+import java.util.ArrayList;
+
+class ScannerResolver {
+    private static final org.slf4j.Logger LOGGER = org.slf4j.LoggerFactory.getLogger(ScannerResolver.class);
```
Comment (derkoe) [inline src/main/java/at/porscheinformatik/sonarqube/licensecheck/ScannerResolver.java:15]: please use org.sonar.api.utils.log.Loggers:
`private static final Logger LOGGER = Loggers.get(ScannerResolver.class);`

File: pom.xml [hunk 2]
```diff
@@ -52,6 +52,17 @@
...
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>1.7.27</version>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-simple</artifactId>
+      <version>1.7.27</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (derkoe) [inline pom.xml:65]: Pleas user SonarQube provided logger API

File: src/main/java/at/porscheinformatik/sonarqube/licensecheck/gradle/GradleDependencyScanner.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import javax.json.JsonObject;
+import javax.json.JsonReader;
+import java.io.*;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Slf4j
```
Comment (derkoe) [inline src/main/java/at/porscheinformatik/sonarqube/licensecheck/gradle/GradleDependencyScanner.java:24]: Please use SonarQube API logger

File: pom.xml [hunk 4]
```diff
@@ -103,6 +114,11 @@
       <artifactId>maven-invoker</artifactId>
       <version>2.2</version>
     </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+      <version>1.18.12</version>
+    </dependency>
```
Comment (derkoe) [inline pom.xml:121]: Please do not use Lombok - write getters/setters directly in the code - it's just for one class

File: README.md [hunk 5]
```diff
@@ -19,9 +19,59 @@ This plugin is compatible:
...
+    }
+
+    licenseReport {
+        allowedLicensesFile = new File("$projectDir/src/main/resources/licenses/allowed-licenses.json")
+        renderers = new JsonReportRenderer('license-details.json', true)
+        filters = [new LicenseBundleNormalizer()]
+        configurations = ALL
+    }
+
+**Step 2:** Update _build.gradle_ file with following code for using sonar-qube plugin
```
Comment (derkoe) [inline README.md:44]: sonar-qube -> SonarQube

File: README.md [hunk 6]
```diff
@@ -19,9 +19,59 @@ This plugin is compatible:
...
+    
+    dependencies {
+        implementation "org.sonarsource.scanner.gradle:sonarqube-gradle-plugin:3.0"
+    }
+    
+**Step 3:** run following command  to generate your report _'license-details.json'_ in _build/reports/dependency-license_
+
+    >gradle generateLicenseReport
+    
+**Step 4:** run following command for sonar-qube
```
Comment (derkoe) [inline README.md:68]: sonar-qube -> SonarQube

File: README.md [hunk 7]
```diff
@@ -19,9 +19,59 @@ This plugin is compatible:
...
+
+    >gradle sonarqube
+    
 ## Installation
 
-Put the pre-built jar-file (from release downloads) in the directory `$SONARQUBE_HOME/extensions/plugins` and
+Put the just builted jar-file (from release downloads) in the directory `$SONARQUBE_HOME/extensions/plugins` and
```
Comment (derkoe) [inline README.md:74]: why did you change this (with typo)?

File: src/main/java/at/porscheinformatik/sonarqube/licensecheck/projectLicense/ProjectLicenseSettingsService.java [hunk 8]
```diff
@@ -64,7 +64,7 @@ private void saveSettings(List<ProjectLicense> projectLicenses)
     {
         Collections.sort(projectLicenses);
         String projectLicensesString = ProjectLicense.createString(projectLicenses);
-        persistentSettings.getSettings().setProperty(PROJECT_LICENSE_KEY, projectLicensesString);
+        persistentSettings.saveProperty(PROJECT_LICENSE_KEY, projectLicensesString);
```
[the remark this answers is not in the document]
Comment (chakrapanich) [reply src/main/java/at/porscheinformatik/sonarqube/licensecheck/projectLicense/ProjectLicenseSettingsService.java]: Reverted.

File: src/main/java/at/porscheinformatik/sonarqube/licensecheck/ScannerResolver.java (hunk 1 above)
Comment (chakrapanich) [reply src/main/java/at/porscheinformatik/sonarqube/licensecheck/ScannerResolver.java]: Fixed

File: pom.xml (hunk 2 above)
Comment (chakrapanich) [reply pom.xml]: Fixed

File: src/main/java/at/porscheinformatik/sonarqube/licensecheck/gradle/GradleDependencyScanner.java (hunk 3 above)
Comment (chakrapanich) [reply src/main/java/at/porscheinformatik/sonarqube/licensecheck/gradle/GradleDependencyScanner.java]: Fixed

File: pom.xml (hunk 4 above)
Comment (chakrapanich) [reply pom.xml]: Fixed

File: README.md (hunk 5 above)
Comment (chakrapanich) [reply README.md]: Fixed

File: README.md (hunk 6 above)
Comment (chakrapanich) [reply README.md]: Fixed

File: README.md (hunk 7 above)
Comment (chakrapanich) [reply README.md]: Reverted

**Comment by derkoe** [comment]
There is still one issue: this PR changes the detection of the scanners.

For example if you have this structure
```
root
- pom.xml
- subproject1
  - package.json
```
then the `PackageJsonDependencyScanner` is not invoked since the root has no `package.json`

The current version is handling case. Maybe always call all scanners and skip the current if the wanted files (`pom.xml`, `package.json`, `build/reports/dependency-license/license-details.json` are not present.

**Comment by chakrapanich** [comment]
> There is still one issue: this PR changes the detection of the scanners.
> 
> For example if you have this structure
> 
> ```
> root
> - pom.xml
> - subproject1
>   - package.json
> ```
> 
> then the `PackageJsonDependencyScanner` is not invoked since the root has no `package.json`
> 
> The current version is handling case. Maybe always call all scanners and skip the current if the wanted files (`pom.xml`, `package.json`, `build/reports/dependency-license/license-details.json` are not present.

yes, changed scanner resolve method.
